internal/telemetry/metric_exporter.go

```diff
@@ -0,0 +1,341 @@
+)
+
+// NewMetricExporter initializes a new MetricExporter
+func NewMetricExporter(p modelpb.BatchProcessor, opts ...ConfigOption) *MetricExporter {
+	cfg := newConfig(opts...)
+
+	return &MetricExporter{
+		processor: p,
+
+		metricFilter:        cfg.MetricFilter,
+		temporalitySelector: cfg.TemporalitySelector,
+		aggregationSelector: cfg.AggregationSelector,
+	}
+}
+
+// MetricExporter is an OpenTelemetry metric Reader which retrieves metrics,
+// filters them and emits them to the specified consumer
+type MetricExporter struct {
+	processor modelpb.BatchProcessor
+
+	metricFilter        []string
+	temporalitySelector metric.TemporalitySelector
+	aggregationSelector metric.AggregationSelector
+}
+
+// Temporality returns the Temporality to use for an instrument kind.
+func (e *MetricExporter) Temporality(k metric.InstrumentKind) metricdata.Temporality {
+	return e.temporalitySelector(k)
+}
+
+// Aggregation returns the Aggregation to use for an instrument kind.
+func (e *MetricExporter) Aggregation(k metric.InstrumentKind) aggregation.Aggregation {
+	return e.aggregationSelector(k)
+}
+
+func (e *MetricExporter) Export(ctx context.Context, rm *metricdata.ResourceMetrics) error {
+	batch := modelpb.Batch{}
+	now := time.Now()
+
+	baseEvent := modelpb.APMEvent{
+		Service: &modelpb.Service{
+			Name:     "apm-server",
+			Language: &modelpb.Language{Name: "go"},
+		},
+		Agent: &modelpb.Agent{
+			Name:    "internal",
+			Version: "unknown",
+		},
+		Event: &modelpb.Event{
+			Received: modelpb.FromTime(now),
+		},
+	}
+
+	for _, scopeMetrics := range rm.ScopeMetrics {
+		ms := make(metricsets)
+		for _, sm := range scopeMetrics.Metrics {
+			if e.isMetricFiltered(sm.Name) {
+				continue
+			}
+
+			if err := addMetric(sm, ms); err != nil {
+				return err
+			}
+		}
+
+		for key, ms := range ms {
+			event := baseEvent.CloneVT()
+			event.Timestamp = modelpb.FromTime(key.timestamp)
+			metrs := make([]*modelpb.MetricsetSample, 0, len(ms.samples))
+			for _, s := range ms.samples {
+				metrs = append(metrs, s)
+			}
+			event.Metricset = &modelpb.Metricset{Samples: metrs, Name: "app"}
+			if ms.attributes.Len() > 0 {
+				event.Labels = modelpb.Labels(event.Labels).Clone()
+				event.NumericLabels = modelpb.NumericLabels(event.NumericLabels).Clone()
+
+				iter := ms.attributes.Iter()
+				for iter.Next() {
+					_, kv := iter.IndexedAttribute()
+					setLabel(string(kv.Key), event, kv.Value.Emit())
+				}
+				if len(event.Labels) == 0 {
+					event.Labels = nil
+				}
+				if len(event.NumericLabels) == 0 {
+					event.NumericLabels = nil
+				}
+			}
+
+			batch = append(batch, event)
+		}
+	}
+	if len(batch) == 0 {
+		return nil
+	}
+
+	return e.processor.ProcessBatch(ctx, &batch)
+}
+
+func (e *MetricExporter) isMetricFiltered(n string) bool {
+	for _, v := range e.metricFilter {
+		if v == n {
+			return false
+		}
+	}
+	return true
+}
+
+func addMetric(sm metricdata.Metrics, ms metricsets) error {
```

Comment:
Could you please add a reference to https://github.com/elastic/apm-data/blob/main/input/otlp/metrics.go? And maybe open an issue/add a TODO to find a way to reuse the apm-data code?

Reply:
I have added a reference. However, I don't think we can/want to reuse the apm-data code. While the logic is extremely similar, the structs are different (apm-data uses the collector ones, while this one must use the SDK ones).

Reply:
Yeah I don't mean directly, I mean indirectly by converting to the collector types. As we discussed live, it would add some overhead and may be complex - so remains to be seen whether it's worthwhile. OTOH it would avoid duplicating the logic for non-trivial metric conversions (grouping by dimensions/attributes, histogram conversion).

Reply:
See https://github.com/elastic/apm-data/issues/144